3.8/graphs-custom-pregel.md

```diff
@@ -0,0 +1,1511 @@
+All other values are considered true.
+
+### Special forms
+
+A _special form_ is special in the sense that it does not necessarily evaluate
+its parameters.
+
+#### `let` statement
+
+_binding values to variables_
+
+```js
+["let", [[var, value]...], expr...]
+```
+
+Expects as first parameter a list of name-value-pairs. Both members of each
+pair are evaluated. `first` has to evaluate to a string. The following
+expression are then evaluated in a context where the named variables are
+assigned to their given values. When evaluating the expression `let` behaves
+like `seq`.
+
+```js
+> ["let", [["x", 12], ["y", 5]], ["+", ["var-ref", "x"], ["var-ref", "y"]]]
+ = 17
+```
+
+#### `seq` statement
+
+_sequence of commands_
+
+```js
+["seq", expr ...]
+```
+
+`seq` evaluates `expr` in order. The result values it the result value of the
+last expression. An empty `seq` evaluates to `none`.
+
+```js
+> ["seq", ["print", "Hello World!"], 2, 3]
+Hello World!
+ = 3
+```
+
+#### `if` statement
+
+_classical if-elseif-else-statement_
+
+```js
+["if", [cond, body], ...]
+```
+
+Takes pairs `[cond, body]` of conditions `cond` and expression `body` and
+evaluates the first `body` for which `cond` evaluates to a value that is
+considered true. It does not evaluate the other `cond`s. If no condition
+matches, it evaluates to `none`. To simulate a `else` statement, set the
+last condition to `true`.
+
+```js
+> ["if", [
+        ["lt?", ["var-ref", "x"], 0],
+        ["-", ["var-ref", "x"]]
+    ], [
+        true, // else
+        ["var-ref", "x"]
+    ]]
+ = 5
+```
+
+#### `match` statement
+
+_not-so-classical switch-statement_
+
+```js
+["match", proto, [c, body]...]
+```
+
+First evaluates `proto`, then evaluates each `c` until `["eq?", val, c]` is
+considered true. Then the corresponding `body` is evaluated and its return
+value is returned. If no branch matches, `none` is returned. This is a C-like
+`switch` statement except that its `case`-values are not treated as constants.
+
+```js
+> ["match", 5,
+  [1, ["A"]],
+  [2, ["B"]],
+  [3, ["C"]],
+  [4, ["D"]],
+  [5, ["E"]]
+]
+  = "E"
+```
+
+#### `for-each` statement
+
+```js
+["for-each", [[var, list]...] expr...]
+```
+
+Behaves similar to `let` but expects a list as `value` for each variable.
+It then produces the cartesian product of all lists and evaluates its
+expression for each n-tuple. The return value is always `none`. The order
+is guaranteed to be reversed lexicographic order.
+
+```js
+> ["for-each", [["x", [1, 2]], ["y", [3, 4]]], ["print", ["var-ref", "x"], ["var-ref", "y"]]]
+1 3
+1 4
+2 3
+2 4
+(no value)
+```
+
+#### `quote` and `quote-splice` statements
+
+_escape sequences for lisp_
+
+```js
+["quote", expr]
+["quote-splice", expr]
+```
+
+`quote`/`quote-splice` copies/splices its parameter verbatim into its output.
+`quote-splice` fails if it is called in a context where it can not splice into
+something, for example at top-level.
```

Comment:
Did you get it from the examples?

Reply:
Yes, the examples are good. But maybe add something like
```suggestion
`quote`/`quote-splice` copies/splices its parameter verbatim into its output, without evaluating them.
`quote-splice` fails if it is called in a context where it can not splice into
something, for example at top-level.
```